File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/internal/interceptors/HeaderInterceptor.java
package com.algolia.internal.interceptors;

import java.io.IOException;
import java.util.Map;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

public final class HeaderInterceptor implements Interceptor {

  private final Map<String, String> headers;

  public HeaderInterceptor(Map<String, String> headers) {
    this.headers = headers;
  }

  @NotNull
  @Override
  public Response intercept(Chain chain) throws IOException {
    okhttp3.Request originalRequest = chain.request();
    okhttp3.Request.Builder builder = originalRequest.newBuilder();

    Headers originalHeaders = originalRequest.headers();
    for (Map.Entry<String, String> header : headers.entrySet()) {
      String key = header.getKey().toLowerCase();
      if (originalHeaders.get(key) != null) {
        builder.header(key, header.getValue());
      }
    }

    okhttp3.Request newRequest = builder.build();
    return chain.proceed(newRequest);
  }
}

File: clients/algoliasearch-client-scala/src/main/scala/algoliasearch/internal/interceptor/HeaderInterceptor.scala
package algoliasearch.internal.interceptor

import okhttp3.{Interceptor, Request, Response}

/** Interceptor that adds headers to a request.
  *
  * @param headers
  *   headers to add to the request
  */
private[internal] class HeaderInterceptor(headers: Map[String, String]) extends Interceptor {

  private val immutableHeaders: Map[String, String] = Map(headers.toSeq: _*)

  override def intercept(chain: Interceptor.Chain): Response = {
    val request: Request = chain.request()
    val builder: Request.Builder = request.newBuilder()
    val requestHeaders = request.headers()

    immutableHeaders.foreach { case (key, value) =>
      if (requestHeaders.get(key) == null) {
        builder.header(key, value)
      }
    }

    val newRequest: Request = builder.build()
    chain.proceed(newRequest)
  }
}
